processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -31,25 +38,180 @@
- *
+ * <p>
  * V1 Storage Format:
- *
+ * <p>
  * byte 1: version (0x1)
- * byte 2 == 0x1 =&gt; allowReverseLookup
- * bytes 3-6 =&gt; numBytesUsed
- * bytes 7-10 =&gt; numElements
+ * byte 2 == 0x1 =>; allowReverseLookup
+ * bytes 3-6 =>; numBytesUsed
+ * bytes 7-10 =>; numElements
  * bytes 10-((numElements * 4) + 10): integers representing *end* offsets of byte serialized values
  * bytes ((numElements * 4) + 10)-(numBytesUsed + 2): 4-byte integer representing length of value, followed by bytes for value
+ * <p>
+ * V2 Storage Format
+ * Meta, header and value files are separate.
+ * Meta File:
+ * byte 1: version (0x2)
+ * byte 2 == 0x1 =>; allowReverseLookup
+ * bytes 3-6: numberOfElementsPerValueFile expressed as power of 2. That means all the value files contains same number of items
+ * except last value file and may have fewer elements.
+ * bytes 7-10 =>; numBytesUsed
+ * bytes 11-14 =>; numElements
+ * bytes 15-18 =>; columnNameLength.
+ * bytes 19-columnNameLength =>; columnName
+ * <p>
+ * Header file name is identified as:  String.format("%s_header", columnName)
+ * value files are identified as: String.format("%s_value_%d", columnName, fileNumber)
+ * number of value files == numElements/numberOfElementsPerValueFile
  */
 public class GenericIndexed<T> implements Indexed<T>
 {
-  private static final byte version = 0x1;
+  private static final byte VERSION_ONE = 0x1;
+  private static final byte VERSION_TWO = 0x2;
+  private final static Ordering ORDERING = Ordering.natural().nullsFirst();
+
+  public static final ObjectStrategy<String> STRING_STRATEGY = new CacheableObjectStrategy<String>()
+  {
+    @Override
+    public Class<? extends String> getClazz()
+    {
+      return String.class;
+    }
+
+    @Override
+    public String fromByteBuffer(final ByteBuffer buffer, final int numBytes)
+    {
+      return StringUtils.fromUtf8(buffer, numBytes);
+    }
+
+    @Override
+    public byte[] toBytes(String val)
+    {
+      if (val == null) {
+        return ByteArrays.EMPTY_ARRAY;
+      }
+      return StringUtils.toUtf8(val);
+    }
+
+    @Override
+    public int compare(String o1, String o2)
+    {
+      return ORDERING.compare(o1, o2);
+    }
+  };
 
-  private int indexOffset;
+  private final ObjectStrategy<T> strategy;
+  private final boolean allowReverseLookup;
+  private final int size;
+  private final BufferIndexed bufferIndexed;
+  private final List<ByteBuffer> valueBuffers;
+  private final ByteBuffer headerBuffer;
+  private final int numElementsPerValueFile;
+
+  private volatile ByteBuffer theBuffer;
+
+  GenericIndexed(
+      ByteBuffer buffer,
+      ObjectStrategy<T> strategy,
+      boolean allowReverseLookup
+  )
+  {
+    this.theBuffer = buffer;
+    this.strategy = strategy;
+    this.allowReverseLookup = allowReverseLookup;
+    size = theBuffer.getInt();
+    numElementsPerValueFile = size;
+
+    int indexOffset = theBuffer.position();
+    int valuesOffset = theBuffer.position() + (size << 2);
```

Comment:
I think no, Hotspot JIT (if not javac) replaces multiplication or division by a static compile-time constant, which is a power of 2, with shift.

Even if it wasn't so, this constructor is not a hot place where every CPU cycle counts